feat(carousel): stay paused after calling `pause()` until `cycle()` is called

divdavem opened the pull request:
After setting the carousel on pause through its `pause()` method and clicking on navigation arrows, the carousel no longer automatically restarts. Calling the `cycle()` method is needed to undo the effect of the `pause()` method.
When `pauseOnHover` is true, hovering on the carousel no longer uses the `pause()` and `cycle()` methods. The hover status is stored separately, so that calling `pause()` and removing the mouse from the carousel no longer restarts it automatically.

The slide event now has a "source" property allowing to know what triggered the event: "timer", "arrowLeft", "arrowRight" or "indicator".
The "pause" boolean status is now also included in the slide event.

Closes #3188

Before submitting a pull request, please make sure you have at least performed the following:

 - [x] read and followed the [CONTRIBUTING.md](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/CONTRIBUTING.md) guide.
 - [x] built and tested the changes locally.
 - [x] added/updated any applicable tests.
 - [x] added/updated any applicable API documentation.
 - [x] added/updated any applicable demos.

benouat commented on src/carousel/carousel.ts:
```diff
@@ -147,18 +165,39 @@ export class NgbCarousel implements AfterContentChecked,
     this.showNavigationIndicators = config.showNavigationIndicators;
   }
 
+  @HostListener('mouseenter')
```
Could you please move back these 2 `@HostListener` to the decorator's metadata ? When there is only one statement in the function, we usually put them directly up there.
→ `'(mouseenter)': '_mouseHover$.next(true)'`

benouat commented on src/carousel/carousel.ts:
```diff
@@ -147,18 +165,39 @@ export class NgbCarousel implements AfterContentChecked,
+    this._mouseHover$.next(true);
+  }
+
+  @HostListener('mouseleave')
```
→ `'(mouseleave)': '_mouseHover$.next(false)'`

benouat commented on src/carousel/carousel.ts:
```diff
@@ -267,6 +306,14 @@ export interface NgbSlideEvent {
+  /**
+   * Whether the pause() method was called (and no cycle() call was done afterwards).
+   */
+  pause: boolean;
```
As @ymeine said. 👍 for renaming to `paused`

benouat commented on src/carousel/carousel.ts:
```diff
@@ -277,4 +324,11 @@ export enum NgbSlideEventDirection {
 }
 
+export enum NgbSlideEventSource {
+  TIMER = 'timer',
```
Do we need to add `<any>` in here ? Such as in `NgbSlideEventDirection`

benouat commented on src/carousel/carousel.ts:
```diff
@@ -267,6 +306,14 @@ export interface NgbSlideEvent {
+   */
+  pause: boolean;
+
+  /** Source triggering the slide change event */
```
I would make the comment a bit more explicative. It's our only way to document `NgbSlideEventSource` as we don't take into account enum in the doc generator.

divdavem commented:
@ymeine and @benouat Thank you for your review.

I have renamed the `pause` property to `paused` in `NgbSlideEvent`.

I have also added the list of possible values in the documentation for the `source` property (as it is done for the `direction` property).

I did not move the content of the `mouseEnter` and `mouseLeave` functions to the `host` property because it would not compile properly in AOT mode because `_mouseHover$` is private.

I did not add `<any>` in the `NgbSlideEventSource` enum because I don't know why it was done for the other enum in the file. For consistency, as you suggested @benouat, it would probably be better to remove `<any>` from the `NgbSlideEventDirection` enum in another PR.
